Added a script for running a Coverity scan.

mbrukman opened the pull request:
This script will help address issue https://github.com/JanusGraph/janusgraph/issues/466 by running Coverity scans directly via our own script rather than the Travis CI plugin.

sjudeng commented on analysis/coverity-scan.sh:
```diff
@@ -0,0 +1,105 @@
+# https://stackoverflow.com/questions/15461737/how-to-execute-xpath-one-liners-from-shell
+# don't work, because Maven's `pom.xml` files use namespaces.
+declare -r PROJECT_VERSION="$(xmllint --xpath '/*[local-name()="project"]/*[local-name()="version"]/text()' pom.xml)"
+declare -r COVERITY_VERSION="${PROJECT_VERSION}/$(date +'%Y-%m-%d')"
```
How about `%Y-%m-%dT%H:%M:%S` to support multiple builds per day? We are technically allowed 3 by project quota, though we haven't needed this so far.

mbrukman replied:
Done.

sjudeng commented on analysis/coverity-scan.sh:
```diff
@@ -0,0 +1,105 @@
+  --post-data "token=${COVERITY_TOKEN}&project=${COVERITY_PROJECT}" \
+  -O "${COVERITY_TOOL_TAR_GZ}" -q \
+  https://scan.coverity.com/download/linux64
+tar zxf "${COVERITY_TOOL_TAR_GZ}"
```
Is it worth stripping out the outer directory and unpacking to generic directory instead (e.g. `cov-analysis-linux64`)? Either way it would be nice to add a check for directory existence before re-downloading? Wouldn't support multiple versions but I don't think that's supported here as it is anyway.

mbrukman replied:
Done.

sjudeng commented on analysis/coverity-scan.sh:
```diff
@@ -0,0 +1,105 @@
+declare -r PROJECT_VERSION="$(xmllint --xpath '/*[local-name()="project"]/*[local-name()="version"]/text()' pom.xml)"
+declare -r COVERITY_VERSION="${PROJECT_VERSION}/$(date +'%Y-%m-%d')"
+
+declare -r COVERITY_DESCRIPTION="Automatic build upload via script"
```
Should we try to capture commit SHA and include in the description/version?

mbrukman replied:
Added git hash to the description.

sjudeng commented on analysis/coverity-scan.sh:
```diff
@@ -0,0 +1,105 @@
+# Run build
+cov-analysis-linux64-*/bin/cov-build \
+  --dir "${COVERITY_BUILD_OUTPUT_DIR}" \
+  mvn -DskipTests=true -T 1C -B -V -pl janusgraph-all -am
```
Missing goal (e.g. `package`). Also can this just be a full build (e.g. remove `-pl janusgraph-all -am`)? I did a full build for the last scan I uploaded and it seems complete. The previous scan result was missing classes but that could be a fluke.

mbrukman replied:
Removed `-pl janusgraph-all -am`.

mbrukman replied:
Added `package` goal.

sjudeng commented on analysis/coverity-scan.sh:
```diff
@@ -0,0 +1,105 @@
+  mvn -DskipTests=true -T 1C -B -V -pl janusgraph-all -am
+
+# Create archive for uploading
+tar cjf "${COVERITY_TAR_BZ2}" "${COVERITY_BUILD_OUTPUT_DIR}"
```
gzip (`czf`)

mbrukman replied:
Done.

mbrukman commented:
I'm running my script now, fixing a few minor issues as they come up during actual run; don't merge yet.

sjudeng commented:
I'm doing same (I added some more feedback above). I've run it a couple times and keep getting coverage round 68%, whereas previously I got 85-95%. I don't think it's the script/mvn call.

Also note if you're running make sure to rebase on master. Otherwise published scan delta might be confusing.

mbrukman commented on analysis/coverity-scan.sh:
```diff
@@ -0,0 +1,105 @@
+# Required to be `cov-int` by Coverity docs:
+# https://scan.coverity.com/download
+declare -r COVERITY_BUILD_OUTPUT_DIR="cov-int"
+declare -r COVERITY_TAR_BZ2="cov-int.tar.bz2"
```
Done.

mbrukman commented on analysis/coverity-scan.sh:
```diff
@@ -0,0 +1,105 @@
+#
+#     $ cd janusgraph
+#     $ env COVERITY_TOKEN="..." COVERITY_EMAIL="..." \
+#           path/to/coverity-scan.sh
```
Done.

sjudeng commented:
Not sure what's going on, I ran in a CentOS 7 Docker container and got 95% but ran twice on an AWS instance (also CentOS 7) and got 68% both times. I'll keep digging. I'm just running locally ... not submitting, so I don't think that counts against the quota, in case you want to test submission yourself.

mbrukman commented on analysis/coverity-scan.sh:
```diff
@@ -0,0 +1,105 @@
+declare -r COVERITY_PROJECT="${GITHUB_ORG}%2F${GITHUB_PROJECT}"
+
+# Download the Coverity toolchain.
+wget \
```
Done.

mbrukman commented:
I believe I've addressed all review comments and pushed the latest version of the script. Now I'm re-running it end-to-end from scratch, rebased on latest `master` change, but note that it will pick up the git hash of the branch, not `master` since it picks up the revision of `HEAD`.

In the future, once this is merged, picking up `HEAD` will do the right thing as it will run from `master`.

mbrukman commented:
Just re-ran on Ubuntu 17.04, got 62% according to the output in `cov-int/build-log.txt`.

mbrukman commented:
I also added a `mvn clean` step before `mvn package` because re-running in the same workspace got me an empty output since there was nothing to do, and the Coverity analysis tool didn't get to see anything.

@sjudeng — should I add that to this PR, or do you think it's unnecessary?

sjudeng commented:
Yeah I think it makes sense ... I got an empty output at one point too, that must have been the reason. I'd add it before (not part of) the `cov-build` call though.

mbrukman commented:
Yep, I added `mvn clean` as a separate command before the Coverity scan to avoid confusing the Coverity tool with that. Let me add that, I'll update the PR.

mbrukman commented:
OK, pushed a few updates:
* added `mvn clean`
* added `-s` flag to the `curl` upload command
* added Coverity-related stuff to `.gitignore` to remove noise

This is the version of the script that I did my last run with; it was submitted to Coverity but probably will fail again due to having < 85% coverage, so our Coverity badge is now red until we fix that.

@sjudeng — please take another look?

sjudeng commented on analysis/coverity-scan.sh:
```diff
@@ -0,0 +1,124 @@
+
+if ! [ -d "${COVERITY_BUILD_OUTPUT_DIR}" ]; then
+  echo "Running Maven build with Coverity analysis ..."
+  mvn -DskipTests=true -T 12C -B -V clean
```
Did you mean `1C`?

mbrukman replied:
I figured, why not clean faster? :-)

If you prefer `1C` I can change it to that, or perhaps scaled by number of CPUs.

mbrukman replied:
Done.

mbrukman commented:
@sjudeng — OK to merge, or are you looking for another reviewer to look over the PR? In which case, care to tag someone? :-)

sjudeng commented:
I'll run through one more time (and hopefully get a more complete scan to upload).

sjudeng commented:
Worked perfect. Also I ran in a CentOS 7 Docker container and again got 95%. Coverity is analyzing now so badge should be back to at least orange shortly.

mbrukman commented:
@sjudeng — have you tried running in Ubuntu? I got much lower percentage there. Is this an issue in my environment, or is Ubuntu unsuited for this task and we should remember to always use CentOS 7?

Is your Dockerfile simply the plain CentOS 7 image, or did you make any modifications to it? Does it make sense to publish a Dockerfile and/or script for running it.

sjudeng commented:
@mbrukman I haven't tried Ubuntu but I've tried on a CentOS 7 AWS instance and I get the same lower coverage there as well.

The base Docker image is just the plain CentOS 7 image. But I just noticed I am setting process/file open limits in the associated Compose file (see [this gist](https://gist.github.com/sjudeng/0c6be52c4a6be5dade6c403a40537e2a)), so that's one thing to check. Otherwise I agree we might want to have the Docker/Compose files in the project for use in running your script through Docker in Travis.

sjudeng commented:
@mbrukman For the record increasing ulimit settings didn't help. Still not sure why percentage is so much higher under Docker CentOS 7, even compared to a CentOS 7 AWS instance.
